### ISSUE TITLE
[Reproduction] SDAR has smaller ALFWorld gains and reverses against GRPO on WebShop 3B under fixed validation

### ISSUE THREAD

**Issue by zhangdw156** [ZJU-REAL/SDAR#58]
Thank you for releasing SDAR. I am trying to reproduce the Qwen2.5 results under a machine-independent task protocol. This appears related to [#43](https://github.com/ZJU-REAL/SDAR/issues/43), [#51](https://github.com/ZJU-REAL/SDAR/issues/51), and [#57](https://github.com/ZJU-REAL/SDAR/issues/57), but I am opening a separate report because it includes fixed full-panel learning curves, exact fork provenance, and a source-parity audit.

## Provenance

I forked the official repository at [`2b35dd7`](https://github.com/ZJU-REAL/SDAR/commit/2b35dd7968b2a0a7e505ae0e65a556f3bcc51769) into [`zhangdw156/SDAR`](https://github.com/zhangdw156/SDAR/tree/exp/iclr). This baseline is code-equivalent to current official `master` [`d511f04`](https://github.com/ZJU-REAL/SDAR/commit/d511f043afe9199b8576fc99ea52e12d84841f32), whose only later change is a README line. The fork includes the official ALFWorld task-specific skill fix from [PR #36](https://github.com/ZJU-REAL/SDAR/pull/36).

At experiment HEAD [`59c6321`](https://github.com/zhangdw156/SDAR/commit/59c63212beb3ebe52ca39eaf05e8f38e2d26da71), the SDAR objective, `sdar_utils.py`, actor joint loss, advantage code, reward manager, and ALFWorld/WebShop prompts are byte-identical to the official baseline. Local changes are fairness/evaluation, 1.5B and four-GPU launchers, active-row compaction, resource/lifecycle changes, an engine-only vLLM seed fix, and tracking cleanup. The fairness commit is [`6afa89a`](https://github.com/zhangdw156/SDAR/commit/6afa89afb38e9b1c8037676f66b114c6d6d714a3).

## Why I added a fairness layer

The inherited environment path could select different task identities across machines: ALFWorld depended on filesystem discovery order and independent worker-local streams, while WebShop worker seeds could change the full goal/price content represented by a numeric index. Independent validation workers could repeat tasks, omit others, and resume on a different panel. I identified these behaviors through a local source audit and cross-machine checks, then tracked the fixes in the method fork history.

The fairness layer supplies explicit canonical manifests with 3,553 ALFWorld training games, all 140 `val_seen` games, all 134 `val_unseen` games, 6,410 WebShop training goals, and 500 WebShop validation goals. It validates uniqueness and counts, uses stable relative paths, makes WebShop goal construction worker-independent and fingerprint-checked, samples WebShop training tasks without replacement before eight-rollout replication, and evaluates each fixed validation task exactly once in deterministic chunks. The 16 × 8 group contract is unchanged.

This protocol is intentionally controlled but not trajectory-identical to the paper. I have not run a fairness-on/off factorial, so I do not infer that fairness caused the discrepancy.

## Experiment protocol

- Qwen2.5-Instruct 1.5B and 3B; one seed; 150 updates; task batch 16 × group 8.
- Effective SDAR coefficient 0.01, beta 5, learning rate `1e-6`, clip 0.2, reference KL 0.01, one PPO epoch.
- ALFWorld: fixed 140-task `val_seen`; unweighted macro-average of six family success rates.
- WebShop: fixed 500-task panel; exact success.
- One stochastic validation sample per task at temperature 0.4, every five updates.
- GRPO uses the same fairness task pools/evaluator but comes from a separate baseline implementation.
- Four trainer GPUs; at 3B, row-padding divisors differ between SDAR and GRPO because of different effective microbatch/world-size settings.

## Results

Paper Table 1 reports 3B SDAR/GRPO ALFWorld success of `84.4/75.0` and WebShop exact success of `68.0/63.3`. My step-150 endpoints are:

| Scale | ALFWorld SDAR / GRPO | WebShop SDAR / GRPO |
|---|---:|---:|
| 1.5B | 70.8 / 61.9 | 61.0 / 62.8 |
| 3B | 69.4 / 65.9 | 58.6 / 63.6 |

There is no local 7B run, and 1.5B has no paper counterpart. The WebShop 3B reversal is late rather than a complete training collapse: normalized AUC is nearly tied (`33.39` SDAR versus `33.90` GRPO), and SDAR reaches 61.6 at update 140 before ending at 58.6.

## Details I may be interpreting incorrectly

1. Paper Algorithm 1 reads as one normalized outcome advantage per trajectory, while the released implementation repeats episode reward onto turn rows and normalizes over those rows by default; trajectory length therefore affects sample mass.
2. Divisibility padding can duplicate randomly selected turn rows. At 3B my SDAR and GRPO jobs used different padding divisors, so the four-GPU resource adaptation may not be numerically neutral.
3. The privileged skill is prepended and the teacher prompt uses unconditional left truncation; I did not archive a direct skill-retention/coverage metric for long WebShop prompts.
4. The paper's 128-task WebShop panel is not identified in the repository; my 500-task panel deliberately differs.

## Questions

1. Could you share the exact Qwen2.5-3B commands, task IDs, model/tokenizer revisions, package versions, random seeds, and checkpoint-selection rule used for Table 1?
2. Is the released turn-row-weighted GRPO behavior the intended implementation of Algorithm 1, or should each trajectory contribute once to group normalization?
3. How should padding rows be weighted, and are there microbatch/padding settings required to reproduce the reported SDAR gains?
4. Are paper results final-update, best-validation, or averaged across checkpoints/seeds?
5. Is there a known WebShop skill-coverage or truncation diagnostic that I should check, especially for the 3B late reversal?
6. Given the similar reports in #43/#51/#57, is there a current known-good reproduction commit or checkpoint you recommend?

I can provide resolved configurations, scalar CSVs, logs, and the complete source-diff audit. Any correction or minimal diagnostic would be greatly appreciated.

**Comment by zhangdw156** [comment]
<img width="2517" height="1714" alt="Image" src="https://github.com/user-attachments/assets/355b8473-8c84-401a-a5dc-92e93a5c1606" />
